sw/device/lib/dif/dif_usbdev.c

```diff
@@ -0,0 +1,934 @@
+ * - Bit 31: True if reading bytes of an incoming packet.
+ *
+ * Note: The first parameters of the functions below are `dif_usbdev_t *` in an
+ * attempt to avoid confusion in cases where the second parameter is a
+ * `uint32_t`.
+ */
+
+static void buffer_op_init(dif_usbdev_t *usbdev) {
+  usbdev->active_buffer_ops = 0;
+}
+
+static void buffer_op_start_write(dif_usbdev_t *usbdev, uint32_t endpoint) {
+  usbdev->active_buffer_ops |= (1 << kEndpointHwInfos[endpoint].bit_index);
+}
+
+static void buffer_op_stop_write(dif_usbdev_t *usbdev, uint32_t endpoint) {
+  usbdev->active_buffer_ops &= ~(1 << kEndpointHwInfos[endpoint].bit_index);
+}
+
+static bool buffer_op_is_writing(dif_usbdev_t *usbdev, uint32_t endpoint) {
+  return (usbdev->active_buffer_ops &
+          (1 << kEndpointHwInfos[endpoint].bit_index)) > 0;
+}
+
+static void buffer_op_start_read(dif_usbdev_t *usbdev) {
+  usbdev->active_buffer_ops |= (1 << USBDEV_READ_BUFFER_OP_BIT_INDEX);
+}
+
+static void buffer_op_stop_read(dif_usbdev_t *usbdev) {
+  usbdev->active_buffer_ops &= ~(1 << USBDEV_READ_BUFFER_OP_BIT_INDEX);
+}
+
+static bool buffer_op_is_reading(dif_usbdev_t *usbdev) {
+  return (usbdev->active_buffer_ops & (1 << USBDEV_READ_BUFFER_OP_BIT_INDEX)) >
+         0;
+}
+
+/**
+ * Utility functions
+ */
+
+static bool is_valid_toggle(dif_usbdev_toggle_t val) {
+  return val == kDifUsbdevToggleEnable || val == kDifUsbdevToggleDisable;
+}
+
+static bool is_valid_power_sense_override(
+    dif_usbdev_power_sense_override_t val) {
+  return val == kDifUsbdevPowerSenseOverrideDisabled ||
+         val == kDifUsbdevPowerSenseOverridePresent ||
+         val == kDifUsbdevPowerSenseOverrideNotPresent;
+}
+
+static dif_usbdev_result_t endpoint_functionality_enable(
+    dif_usbdev_t *usbdev, uint32_t reg_offset, uint32_t endpoint,
+    dif_usbdev_toggle_t new_state) {
+  if (usbdev == NULL) {
+    return kDifUsbdevResultInvalidArgument;
+  }
+
+  if (!is_valid_endpoint(endpoint) || !is_valid_toggle(new_state)) {
+    return kDifUsbdevResultGeneralError;
+  }
+
+  if (kDifUsbdevToggleEnable) {
+    mmio_region_nonatomic_set_bit32(usbdev->base_addr, reg_offset,
+                                    kEndpointHwInfos[endpoint].bit_index);
+  } else {
+    mmio_region_nonatomic_clear_bit32(usbdev->base_addr, reg_offset,
+                                      kEndpointHwInfos[endpoint].bit_index);
+  }
+
+  return kDifUsbdevResultOK;
+}
+
+/**
+ * USBDEV DIF library functions.
+ */
+
+dif_usbdev_result_t dif_usbdev_init(dif_usbdev_config_t *config,
+                                    dif_usbdev_t *usbdev) {
+  if (usbdev == NULL || config == NULL) {
+    return kDifUsbdevResultInvalidArgument;
+  }
+
+  // Check enum fields
+  if (!is_valid_toggle(config->differential_rx) ||
+      !is_valid_toggle(config->differential_tx) ||
+      !is_valid_toggle(config->single_bit_eop) ||
+      !is_valid_power_sense_override(config->power_sense_override)) {
+    return kDifUsbdevResultGeneralError;
```

Comment:
I think we were only doing this kind of checking in `switch` statements, right? @lenary could you clarify?

Reply:
I included this check because an equivalent `switch` would have it but wrote it in this form to make this cheks earlier. @lenary @mcy please let me know if you want me to make any changes.

Reply:
From the style guide:
> DIFs must ensure, if they only accept a subset of an enum, that the argument is within that subset. However, DIFs may assume, for checking preconditions, that any enum argument is one of the enum constants.

You're accepting all sates, so you don't have to check `is_valid_toggle`, or `is_valid_power_sense_override`. I'm happy that what you're doing if the values are not one of the enum values is "safe" in the implementation of this function.